crates/sqllib/src/dec.rs

```diff
@@ -0,0 +1,923 @@
+        }
+    }
+}
+
+impl HasZero for SqlDecimal {
+    #[inline]
+    fn is_zero(&self) -> bool {
+        Zero::is_zero(self)
+    }
+
+    #[inline]
+    fn zero() -> Self {
+        Zero::zero()
+    }
+}
+
+impl HasOne for SqlDecimal {
+    #[inline]
+    fn one() -> Self {
+        One::one()
+    }
+}
+
+impl TryInto<i8> for SqlDecimal {
+    type Error = TryFromDecimalError;
+
+    fn try_into(self) -> Result<i8, Self::Error> {
+        let mut context = CONTEXT_PROTO.clone();
+        context.set_rounding(Rounding::Down);
+        let mut val = self.value;
+        context.round(&mut val);
+        context.clear_status();
+        context.try_into_i8(val)
+    }
+}
+
+impl TryInto<i16> for SqlDecimal {
+    type Error = TryFromDecimalError;
+
+    fn try_into(self) -> Result<i16, Self::Error> {
+        let mut context = CONTEXT_PROTO.clone();
+        context.set_rounding(Rounding::Down);
+        let mut val = self.value;
+        context.round(&mut val);
+        context.clear_status();
+        context.try_into_i16(val)
+    }
+}
+
+impl TryInto<i32> for SqlDecimal {
+    type Error = TryFromDecimalError;
+
+    fn try_into(self) -> Result<i32, Self::Error> {
+        let mut context = CONTEXT_PROTO.clone();
+        context.set_rounding(Rounding::Down);
+        let mut val = self.value;
+        context.round(&mut val);
+        context.clear_status();
+        context.try_into_i32(val)
+    }
+}
+
+impl TryInto<i64> for SqlDecimal {
+    type Error = TryFromDecimalError;
+
+    fn try_into(self) -> Result<i64, Self::Error> {
+        let mut context = CONTEXT_PROTO.clone();
+        context.set_rounding(Rounding::Down);
+        let mut val = self.value;
+        context.round(&mut val);
+        context.clear_status();
+        context.try_into_i64(val)
+    }
+}
+
+impl TryInto<f32> for SqlDecimal {
+    type Error = TryFromDecimalError;
+
+    fn try_into(self) -> Result<f32, Self::Error> {
+        let mut context = CONTEXT_PROTO.clone();
+        context.set_rounding(Rounding::Down);
+        let mut val = self.value;
+        context.round(&mut val);
+        context.try_into_f32(val)
+    }
+}
+
+impl TryInto<f64> for SqlDecimal {
+    type Error = TryFromDecimalError;
+
+    fn try_into(self) -> Result<f64, Self::Error> {
+        let mut context = CONTEXT_PROTO.clone();
+        context.set_rounding(Rounding::Down);
+        let mut val = self.value;
+        context.round(&mut val);
+        context.try_into_f64(val)
+    }
+}
+
+impl Eq for SqlDecimal {}
+
+impl PartialOrd for SqlDecimal {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for SqlDecimal {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        let mut context = CONTEXT_PROTO.clone();
+        // This returns Option for NaN, Inf, and other illegal values
+        context.partial_cmp(&self.value, &other.value).unwrap()
```

Comment:
The `unwrap` here seems scary. I can panic just by comparing a NaN?

Reply:
We don't support NaN or Inf for DECIMAL, so this should be a panic.